### PR TITLE
add 8 "push" lemmas for `nth_error`

### PR DESCRIPTION
**Pull request by andres-erbsen**
- [x] Added **changelog**.

Here are 8 standard lemmas characterizing existing list definitions in terms of `nth_error`.

Cc @Villetaneuse

### CODE REVIEW

File: theories/Lists/List.v
Comment (Villetaneuse) [inline theories/Lists/List.v:659]: I think it would be better to use 0 (zero) consistently here.

File: theories/Lists/List.v
Comment (Villetaneuse) [inline theories/Lists/List.v:662]: I would prefer "succ" instead of "S". Now, I know that "S" is the "local convention in this file", but maybe we could also change some names (there are not that many) and give notations for compatibility (with or without deprecation?) for lemmas using S for succ and O (oh) for 0 (zero)?

**Review by Villetaneuse — COMMENTED** [review]
I think these lemmas are without a doubt very useful and should be there.

Last comment about proofs. Is there a style guide somewhere for new proofs in the stdlib? For instance is `destruct` prefered to `case`, is it encouraged to use automation?
I think this should be clarified (before or after merging this PR, I don't know).

File: theories/Lists/List.v
Comment (andres-erbsen) [reply theories/Lists/List.v:662]: I have no opinion here; I don't think there is any explicit guideline to follow.

File: theories/Lists/List.v
Comment (andres-erbsen) [reply theories/Lists/List.v:659]: I think it would be overall better if nat used succ rather than S, but changing it in just this one new lemma might be even more confusing. `Search "_S"` several useful examples.

File: theories/Lists/List.v
Comment (Villetaneuse) [reply theories/Lists/List.v:662]: I have checked what I get in `Arith` and `List` (this already loads 128 files) :
```coq
Require Import Arith List.

(* Coq.ZArith.Zeven, deprecated (informally)
   TODO: deprecate this library file *)
About Zeven.Zodd_Sn.
About Zeven.Zeven_Sn.

(* In Arith.PeanoNat : *)
About Nat.Odd_alt_S. (* Constructors of mutually inductive props *)
About Nat.Even_alt_S. (* Constructors of mutually inductive props *)
About Nat.EvenT_S_OddT. (* TODO: add new name *)
About Nat.OddT_S_EvenT. (* TODO: add new name *)
About Nat.double_S. (* TODO: add new name *)
About PeanoNat.lt_S_n. (* with register, probably for backwards compatibility
  TODO: add synonoym, there is an equivalence Nat.succ_lt_mono *)
About PeanoNat.lt_n_Sm_le. (* Same, one side of Nat.lt_succ_r *)
About PeanoNat.le_lt_n_Sm. (* Same *)

(* In Init.Peano *)
About n_Sn. (* has synonym Nat.neq_succ_diag_r *)
About pred_Sn. (* has synonym Nat.pred_succ *)
About le_S. (* has synonym Nat.le_le_succ_r *)
About O_S. (* has synonym Nat.neq_0_succ *)
About le_n_S. (* TODO: add synonym, there is an equivalence Nat.succ_le_mono *)
About le_S_n. (* Same *)
About eq_add_S. (* has synonym Nat.succ_inj *)
About eq_S.
About not_eq_S.
About plus_n_Sm.
About plus_Sn_m.
About mult_n_Sm.

(* In Arith.Between, I have no idea how this file can be useful
   TODO: investigate deprecation *)
About in_int_Sp_q. 
About in_int_S.
About exists_S_le.
About exists_le_S.
About exists_S.
About bet_S.
About in_int_p_Sq.

(* Why I get BinNat with Arith is a mystery to me. *)
About BinNat.Nmult_Sn_m. (* flagged as "compatibility lemma" *)

(* In Arith.Compare_dec
   TODO: Check if there are alternatives to Compare_dec *)
About le_le_S_dec.
(* In Arith.Peano_dec
   TODO: Check if there are alternatives to Peano_dec *)
About O_or_S.

(* In List *)
About nth_error_S.
About nth_error_cons_S.
About nth_S.
About seq_S.
About nth_S_cons.
About nth_error_Some.
```
I think it's fair to say that most lemmas with "_S" are here for compatibility reasons.
Of course, the TODOs are for me, not for this PR.

Comment (andres-erbsen) [reply theories/Lists/List.v]: Ok, I did `_succ` added a deprecated compat notation as you suggested.

**Comment by andres-erbsen** [comment]
@coqbot run full ci

I don't understand why fiat-crypto is not passing, it is passing on my machine with identical last-commit diff in fiat-crypto.

**Comment by SkySkimmer** [comment]
I see stdlib failures not fiat crypto.

**Comment by andres-erbsen** [comment]
@coqbot run full CI

The stdlib failure was this PR getting stale (detected by repoen), still no clue on previous fiat-crypto failures.

**Comment by andres-erbsen** [comment]
@coqbot run full ci

**Comment by andres-erbsen** [comment]
@coqbot run full ci

**Comment by andres-erbsen** [comment]
I clicked rerun but only some of the jobs got restarted...

**Comment by andres-erbsen** [comment]
@coqbot run full ci

**Comment by andres-erbsen** [comment]
@coq/stdlib-maintainers please take a look  
I think the Windows failure is https://github.com/coq/coq/issues/18607

**Comment by andres-erbsen** [comment]
@coq/stdlib-maintainers I would really appreciate a review for this! I am building on top of this PR in Zmod and being able to merge dependencies like this separately would substantially reduce the size of the main PR.

File: theories/Lists/List.v
Comment (proux01) [reply theories/Lists/List.v]: In any case, this should be documented (for instance in some `theories/CONTRIBUTING.md`, otherwise the next contributor may very well decide that "S is better than succ" and we'll never reach any stable state.

**Review by proux01 — APPROVED** [review]
Let's rerun CI to be on the safe side, and merge.

**Comment by proux01** [comment]
@coqbot run full ci

**Comment by andres-erbsen** [comment]
Windows [failure](https://github.com/coq/coq/actions/runs/7747623649/job/21128510545?pr=18563#step:6:1783) is unrelated:
```
# File "ide/coqide/coqide_WIN32.ml.in", line 51, characters 37-68:
# Error: Unbound value Glib.Io.channel_of_descr_socket
```

**Pull request merged**
